### PR TITLE
Color support

### PR DESCRIPTION
**Pull request by johnnyr26**
Adds checkbox that allows/disables color backgrounds
Generates new color if the default colors run out

### CODE REVIEW

**Comment by johnnyr26** [comment]
Fix anything you need to fix but I tried my best to make all changes necessary for the color implementation. I also tried my best to keep the style of the code as similar as it can be to yours.
I haven't tested it yet so there may be some mistakes I'm not aware of.

**Comment by jbis9051** [comment]
Why close?

**Comment by johnnyr26** [comment]
My bad, my mistake

**Comment by johnnyr26** [comment]
I tried to make a change to add semicolons to the lines I forgot to add semicolons, so I thought I needed to close it but I realized I didn't

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ module.exports = (timedScheduleObject) => {
         $('table thead tr').append(`<th class="letter">${day}</th>`);
     });
     const numOfClasses = Object.values(timedScheduleObject)[0].length;
+    const colors = ["tomato", "dodgerblue", "yellow", "orange", "mediumseagreen", "lightblue", "pink", "#d6ff99", "gainsboro", "#FAE7B5", "#a89de1", "burlywood"]
```
Comment (jbis9051) [inline SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js]: Should be globally declared (same array each time)

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js [hunk 2]
```diff
@@ -34,6 +35,18 @@ module.exports = (timedScheduleObject) => {
                     if (theClass.teacher) {
                         classCode += `<span class="teacher">${theClass.teacher}</span>`;
                     }
+                    if (theClass.color) {
+                        if(colors[i]) {
+                            document.getElementsByClassName("class").style.backgroundColor = colors
```
Comment (jbis9051) [inline SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js]: DOM API not available within Node.js

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js [hunk 3]
```diff
@@ -34,6 +35,18 @@ module.exports = (timedScheduleObject) => {
...
+                        } else {
+                            let color = []
+                            for (_= 0;_<3;_++) {
+                                color.push(Math.floor(Math.random() * 255))
+                            }
+                            document.getElementsByClassName("class").style.backgroundColor = "rgba(" + color[0] + "," + color[1] + "," + color[2] + ", .5)"
+                            // In the case that all of the colors are used, the program would generate a new color with a light opacity in order for the class to be readable
```
Comment (jbis9051) [inline SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:41]: We know there is a max amount of classes possible (6 classes a day times 8 days = 48 colors). I suggest we create a bank of 48 colors.

File: views/index.ejs [hunk 4]
```diff
@@ -71,6 +71,7 @@
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="teacher-toggle"><span>Show Teachers</span></label></div>
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="room-toggle"><span>Show Room</span></label></div>
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="free-toggle"><span>Show Free Periods</span></label></div>
+    <div class="checkbox-wrapper"><labeL><input type="checkbox" id="color-toggle"><span>Show Color Background</span></label></div>
```
Comment (jbis9051) [inline views/index.ejs]: Are we not showing the colors by default?

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js [hunk 5]
```diff
@@ -27,7 +28,21 @@ module.exports = (timedScheduleObject) => {
...
+                    if (theClass.color) {
+                        if(colors[i]) {
+                            classCode = `<td class="class" style="background-color: ${colors}">`;
+                        } else {
+                            let colorArray = [];
+                            for (_= 0;_<3;_++) {
+                                color.push(Math.floor(Math.random() * 255));
+                            }
+                            const newColor = `rgba("${colorArray[0]}","${colorArray[1]}", "${colorArray[2]}", .5)`;
+                            classCode = `<td class="class" style="background-color: ${newColor}">`;
```
Comment (jbis9051) [inline SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:40]: How are we gonna disable this with the toggle?

File: public/js/main.js [hunk 6]
```diff
@@ -142,6 +142,10 @@ $('#submit-button').addEventListener('click', e => {
                 $('#free-toggle').setAttribute('disabled','');
                 $('#free-toggle').checked = false;
             }
+            if($$('table .color').length == 0) {
+                $('#color-toggle').setAttribute('disabled.','');
```
Comment (jbis9051) [inline public/js/main.js]: Typo: Extra "." for the "disabled" attribute

File: public/js/main.js [hunk 7]
```diff
@@ -213,6 +217,14 @@ $('#free-toggle').addEventListener('change', evt => {
...
-        $$('table .free-text').forEach(el => el.setAttribute('hidden',''))
+        $$('table .free-text').forEach(el => el.setAttribute('hidden',''));
+    }
+});
+
+$('#color-toggle').addEventListener('change', evt => {
+    if($('#color-toggle').checked){
+        $$('table .color').forEach(el => el.removeAttribute('hidden'));
+    } else {
+        $$('table .color').forEach(el => el.setAttribute('hidden',''));
```
Comment (jbis9051) [inline public/js/main.js]: I don't think there is any elements that will match this selector.

**Review by jbis9051 — CHANGES_REQUESTED** [review]
Please fix changes mentioned. Also I suggest editing the timedScheduleObject creation script to include indices.

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js (hunk 2 above)
Comment (johnnyr26) [reply SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js]: Look at the new file change that occurred after this

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js (hunk 3 above)
Comment (johnnyr26) [reply SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:41]: Hard to do when the colors get really similar

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js (hunk 1 above)
Comment (johnnyr26) [reply SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js]: Okay

File: views/index.ejs (hunk 4 above)
Comment (johnnyr26) [reply views/index.ejs]: We can do that

**Comment by jbis9051** [comment]
> I suggest editing the timedScheduleObject creation script to include indices.

This has been resolved with https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/90c620c04d4848c9dea9d60d348a302bc426f91a. Use the `classNum` to index the color bank.

**Comment by jbis9051** [comment]
I suggest creating an array of 48 css names, instead of an array of color hexes/names. Then creating a CSS style sheet with: 
```CSS
.color .some_class_name {
     background: #some_color;
}
```
for each.  We can then simply toggle the `.color` class. This will allow us to toggle colors without having to keep track of the individual colors. 

Also for the colors themselves, I suggest soft colors that will show well with text.  For example, instead of `blue` maybe `#99badd`

**Comment by johnnyr26** [comment]
I got work to do so I’m done for the day. I’ll talk and work tomorrow

Just one quick help. Don’t know JQuery for this, but can you write me the JQ code that will set the background color back to white

$('#color-toggle').addEventListener('change', evt => {
    if(!$('#color-toggle').checked){
        //Resets the background color to white
    }
});

Thanks.

File: SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js (hunk 3 above)
Comment (jbis9051) [reply SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:41]: You really only need like 10-12 solid colors, after that in can be same base color just a different shade as these will rarely ever come up.

**Comment by jbis9051** [comment]
This is not jQuery, see [L1](https://github.com/jbis9051/Blind-Brook-Calendar-Creator/blob/master/public/js/main.js#L1).

In jQuery it would be:

```javascript
$('#color-toggle').change(evt => {
    if(!$('#color-toggle').checked){ // however dup selector
       $(selector).css('background','white');
    }
});
```
Plain js:
```javascript
document.querySelector('#color-toggle').addEventListener('change', evt => {
    if(!document.querySelector('#color-toggle').checked){ 
       document.querySelectorAll(selector).forEach(el => el.style.background = "white");
    }
});
```

**Comment by jbis9051** [comment]
Please implement the updated color generation system (See: https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/e2ee45d9c8daa6bec2c9b5dc68fe94b50444f6a8). Also please separate https://github.com/jbis9051/Blind-Brook-Calendar-Creator/pull/3/commits/5b9020881ca13824354acbfb083d26817b8ee8c5. I will merge that.

**Comment by jbis9051** [comment]
I suggest getting an IDE. My suggestion is [Jet Brain's WebStorm](https://www.jetbrains.com/webstorm/). You can get a free student account.

**Comment by johnnyr26** [comment]
Can I get a list of all the things you want me to do?
It's hard to see what I have to do when they are all separated

**Comment by jbis9051** [comment]
1. Create a separate PR for https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/5b9020881ca13824354acbfb083d26817b8ee8c5
2. Create a new PR for Color support
3. Using the stuff from https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/90c620c04d4848c9dea9d60d348a302bc426f91a and https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/e2ee45d9c8daa6bec2c9b5dc68fe94b50444f6a8, make the colors and toggle work
**Don't update this PR, create a new one**

Optional but suggested:

1. Get an IDE (Webstorm)

**Comment by jbis9051** [comment]
Color support added with: https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/f264d17d7453af2901d6c1853593768422537952

**Comment by johnnyr26** [comment]
Hey,

I saw you were able to add the color support feature.

I’ll redirect my website to yours and tell all the administrators that are relying on my site once you add my name to the bottom of the site as well as yours like we agreed to.

Let me know if you need anything else.

**Comment by jbis9051** [comment]
@JRProgramming Please see https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/441c723c743ca062d3f2e243ea48eeac1925b8f4. Please close the "Combine" issue on your project once you have redirected your site.

**Comment by johnnyr26** [comment]
My site now redirects to yours.

I’ll let them know when I see them that my url now redirects to your site and to use your url from now.

Glad we were able to combine

**Pull request closed without merging**